activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -584,6 +592,7 @@ def reconnect!(restore_transactions: false)
         retries_available = connection_retries
 
         @lock.synchronize do
+          start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
```

Comment:
It would make sense to move this outside of the `begin/rescue` so that it acts as a true deadline. e.g.

If you have `retries_available = 3` and `retry_deadline = 2.0`

  - first try fail after 1s, we retry
  - second try fail after another 1s, we don't retry a third time because we're past the deadline.

Reply:
You can even do something like (pseudo-code):

```ruby
deadline = retry_deadline && Process.clock_gettime(Process::CLOCK_MONOTONIC) + retry_deadline
```

Then in the rescue:

```ruby
if deadline && deadline < Process.clock_gettime(Process::CLOCK_MONOTONIC) 
  # bail out
end
```

Reply:
Oh yes , definitely doesn't make sense to reset the deadline at the start of every block 🤦‍♀️ On it!